### PR TITLE
Add index.routing.allocation.include._tier_preference setting

### PR DESCRIPTION
**Pull request by dakrone**
This commit adds the `index.routing.allocation.include._tier_preference` setting to the
`DataTierAllocationDecider`. This special-purpose allocation setting lets a user specify a
preference-based list of tiers for an index to be assigned to. For example, if the setting were set
to:

```
"index.routing.allocation.include._tier_preference": "data_hot,data_warm,data_content"
```

If the cluster contains any nodes with the `data_hot` role, the decider will only allow them to be
allocated on the `data_hot` node(s). If there are no `data_hot` nodes, but there are `data_warm` and
`data_content` nodes, then the index will be allowed to be allocated on `data_warm` nodes.

This allows us to specify an index's preference for tier(s) without causing the index to be
unassigned if no nodes of a preferred tier are available.

Subsequent work will change the ILM migration to make additional use of this setting.

Relates to #60848

### CODE REVIEW

**Comment by dakrone** [comment]
Since this setting is not part of the existing preference group, we need to handle backwards compatibility. I think we have a couple of options here:

- Find a way to determine the node versions when the index is created, ignoring the setting for versions that don't support it.
- Collapsed the setting into one of the existing groups, so something like: `index.routing.allocation.include._tier_preference` or `index.routing.allocation.include_prefer_tier`

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/cluster/routing/allocation/DataTierAllocationDecider.java [hunk 1]
```diff
@@ -186,6 +231,29 @@ private Decision shouldClusterFilter(DiscoveryNode node, RoutingAllocation alloc
         OR
     }
 
+    /**
+     * Given a string of comma-separated prioritized tiers (highest priority
+     * first) and an allocation, find the highest priority tier for which nodes
+     * exist. If no nodes for any of the tiers are available, returns an empty
+     * {@code Optional<String>}.
+     */
+    private static Optional<String> preferredTier(String prioritizedTiers, RoutingAllocation allocation) {
```
Comment (andreidan) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/cluster/routing/allocation/DataTierAllocationDecider.java:240]: would this be better named as `preferedAvailableTier` or `preferedPresentTier`?

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/cluster/routing/allocation/DataTierAllocationDecider.java [hunk 2]
```diff
@@ -186,6 +231,29 @@ private Decision shouldClusterFilter(DiscoveryNode node, RoutingAllocation alloc
...
+        String[] tiers = Strings.tokenizeToStringArray(prioritizedTiers, ",");
+        return Arrays.stream(tiers).filter(tier -> tierNodesPresent(tier, allocation)).findFirst();
+    }
+
+    private static boolean tierNodesPresent(String singleTier, RoutingAllocation allocation) {
```
Comment (andreidan) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/cluster/routing/allocation/DataTierAllocationDecider.java:245]: should we make both these methods as default visible and add unit tests?

**Review by andreidan — CHANGES_REQUESTED** [review]
Thanks for working on this Lee. 

I've left a couple of rather minor suggestions. 

Can you please add some tests to show illustrate the include/exclude/require precedence over `preferred`?

**Comment by dakrone** [comment]
Thanks Andrei, I made the changes you suggested

**Comment by dakrone** [comment]
@elasticmachine run elasticsearch-ci/packaging-sample-windows

**Comment by dakrone** [comment]
@elasticmachine update branch

**Comment by dakrone** [comment]
@elasticmachine test this please

**Comment by dakrone** [comment]
@elasticmachine run elasticsearch-ci/2

**Review by andreidan — APPROVED** [review]
LGTM, thanks Lee

**Pull request merged**